fotoapparat/src/main/java/io/fotoapparat/hardware/v1/capabilities/CapabilitiesFactory.java

```diff
@@ -21,69 +22,77 @@
-		HashSet<Flash> result = new HashSet<>();
-
-		for (String flashMode : supportedFlashModes(parameters)) {
-			result.add(
-					FlashCapability.toFlash(flashMode)
-			);
-		}
-
-		return result;
-	}
-
-	@NonNull
-	private List<String> supportedFlashModes(Camera.Parameters parameters) {
-		List<String> supportedFlashModes = parameters.getSupportedFlashModes();
-		return supportedFlashModes != null
-				? supportedFlashModes
-				: Collections.singletonList(Camera.Parameters.FLASH_MODE_OFF);
-	}
-
-	private Set<FocusMode> extractFocusModes(Camera.Parameters parameters) {
-		HashSet<FocusMode> result = new HashSet<>();
-
-		for (String focusMode : parameters.getSupportedFocusModes()) {
-			result.add(
-					FocusCapability.toFocusMode(focusMode)
-			);
-		}
-
-		return result;
-	}
+    /**
+     * @return {@link Capabilities} from given camera parameters.
+     */
+    public Capabilities fromParameters(Camera.Parameters parameters) {
+        return new Capabilities(
+                extractPictureSizes(parameters),
+                extractPreviewSizes(parameters),
+                setPreviewScaleTypes(),
+                extractFocusModes(parameters),
+                extractFlashModes(parameters)
+        );
+    }
+
+    private Set<Size> extractPreviewSizes(Camera.Parameters parameters) {
+        return mapSizes(parameters.getSupportedPreviewSizes());
+    }
+
+    private Set<Size> extractPictureSizes(Camera.Parameters parameters) {
+        return mapSizes(parameters.getSupportedPictureSizes());
+    }
+
+    private Set<Size> mapSizes(Collection<Camera.Size> sizes) {
+        HashSet<Size> result = new HashSet<>();
+
+        for (Camera.Size size : sizes) {
+            result.add(new Size(
+                    size.width,
+                    size.height
+            ));
+        }
+
+        return result;
+    }
+
+    private Set<Flash> extractFlashModes(Camera.Parameters parameters) {
+        HashSet<Flash> result = new HashSet<>();
+
+        for (String flashMode : supportedFlashModes(parameters)) {
+            result.add(
+                    FlashCapability.toFlash(flashMode)
+            );
+        }
+
+        return result;
+    }
+
+    @NonNull
+    private List<String> supportedFlashModes(Camera.Parameters parameters) {
+        List<String> supportedFlashModes = parameters.getSupportedFlashModes();
+        return supportedFlashModes != null
+                ? supportedFlashModes
+                : Collections.singletonList(Camera.Parameters.FLASH_MODE_OFF);
+    }
+
+    private Set<FocusMode> extractFocusModes(Camera.Parameters parameters) {
+        HashSet<FocusMode> result = new HashSet<>();
+
+        for (String focusMode : parameters.getSupportedFocusModes()) {
+            result.add(
+                    FocusCapability.toFocusMode(focusMode)
+            );
+        }
+
+        return result;
+    }
+
+    private Set<ScaleType> setPreviewScaleTypes() {
```

Comment:
As already mentioned - those scale types should not be here. Available scale types can't possibly depend on camera parameters.